crates/storage/provider/src/providers/database/provider.rs

```diff
@@ -251,105 +255,109 @@ impl<TX: DbTxMut + DbTx> DatabaseProvider<TX> {
-                hash_map::Entry::Vacant(entry) => {
-                    let new_storage = plain_storage_cursor
-                        .seek_by_key_subkey(address, old_storage.key)?
-                        .filter(|storage| storage.key == old_storage.key)
-                        .unwrap_or_default();
-                    entry.insert((old_storage.value, new_storage.value));
-                }
-                hash_map::Entry::Occupied(mut entry) => {
-                    entry.get_mut().0 = old_storage.value;
+            if !bundle_builder.get_states().contains(&address) {
+                if let Some(present_info) =
+                    plain_accounts_cursor.seek_exact(address)?.map(|kv| kv.1)
+                {
+                    bundle_builder = bundle_builder
+                        .state_original_account_info(address, into_revm_acc(present_info));
+                    bundle_builder = bundle_builder
+                        .state_present_account_info(address, into_revm_acc(present_info));
                 }
-            };
-
-            reverts
-                .entry(block_number)
-                .or_default()
-                .entry(address)
-                .or_default()
-                .1
-                .push(old_storage);
+            }
+            let new_storage = plain_storage_cursor
+                .seek_by_key_subkey(address, old_storage.key)?
+                .filter(|storage| storage.key == old_storage.key)
+                .unwrap_or_default();
+            bundle_builder = bundle_builder.state_storage(
+                address,
+                HashMapRevm::from([(
+                    new_storage.key.into(),
+                    (old_storage.value, new_storage.value),
+                )]),
+            );
+            // insert old storage into reverts.
+            bundle_builder = bundle_builder.revert_storage(
+                block_number,
+                address,
+                vec![(old_storage.key.into(), old_storage.value)],
+            );
         }
 
+        // build the `BundleState`
+        let bundle_state = bundle_builder.build();
+
         if UNWIND {
             // iterate over local plain state remove all account and all storages.
-            for (address, (old_account, new_account, storage)) in state.iter() {
-                // revert account if needed.
-                if old_account != new_account {
-                    let existing_entry = plain_accounts_cursor.seek_exact(*address)?;
-                    if let Some(account) = old_account {
-                        plain_accounts_cursor.upsert(*address, *account)?;
-                    } else if existing_entry.is_some() {
-                        plain_accounts_cursor.delete_current()?;
-                    }
+            let state_changeset = bundle_state.clone().into_plain_state(OriginalValuesKnown::No);
+            // revert account if needed.
+            for (address, account) in state_changeset.accounts {
```

Comment:
Is this okay, or is there a better way, doesn't the bundle builder have all the needed info so we could skip both the cloning and the `into_plain_state`?

Reply:
I mean by is it okay what happens if previous bundle state does not have account it would be None, but present plain state would contain it (it should remove it)?